### ISSUE TITLE
Add rank-aware dataset sharding to incin-data

### ISSUE THREAD

**Issue by xupremix** [xupremix/incin#98]
> Filed against the `feature_request.yml` template. Category: **Dataset / Dataloader (`incin-data`)**.

## Problem

`incin-data` has no concept of a rank. `grep -ril 'rank\|shard' crates/incin-data/src/` returns nothing.

That makes #97 incorrect by construction. N ranks each build the same `DataLoader` over the same dataset with the same shuffle seed, so every rank draws the same samples. The gradients are then all-reduced into an average of N identical gradients, and the run trains on **one Nth of the data it appears to, at N times the cost**, while every metric looks plausible.

This is a small crate-local change that silently invalidates a large feature if it is missed, which is why it is its own issue rather than a bullet inside #97.

## Proposed API sketch

Loader construction below is written in `incin-data`'s builder style; the exact
current spelling is whatever that crate exposes, and only the `// proposed`
lines are new surface.

```rust
use incin::prelude::*;
use incin_data::{DataLoader, DistributedSampler};

// ── single process: unchanged, byte for byte ──
let loader = DataLoader::new(dataset)
    .batch_size(32)
    .shuffle(Seed(42))
    .build()?;

// ── data-parallel: one added call ──
let loader = DataLoader::new(dataset)
    .batch_size(32)
    .shuffle(Seed(42))
    .sampler(DistributedSampler {                       // proposed
        world_size: context.world_size(),
        rank: context.rank(),
        remainder: RemainderPolicy::Drop,               // proposed: explicit, not implicit
    })
    .build()?;

// Epoch must be advertised so the shuffle differs across epochs while
// staying consistent across ranks, with no inter-rank communication.
for epoch in 0..epochs {
    loader.set_epoch(epoch)?;                            // proposed
    for batch in &loader { /* ... */ }
}
```

`world_size` and `rank` come from the existing `DistributedContext` (`RANK_ENV`, `WORLD_SIZE_ENV` in `crates/incin-core/src/dist/context/`), so this reads identity from the same place the collectives do rather than inventing a second source.

```rust
// proposed
pub enum RemainderPolicy {
    /// Drop the tail so every sample is unique. Ranks may see unequal step counts
    /// unless the dataset divides evenly, so this pairs with `drop_last`.
    Drop,
    /// Repeat samples from the start so every rank runs the same number of steps.
    /// Correct for training; wrong for evaluation, because samples are double-counted.
    Pad,
}
```

## Shape and type-system invariants

No tensor shapes change. The invariants are set-theoretic and they are the whole correctness argument:

| Invariant | Statement |
|---|---|
| **Disjointness** | For ranks `i != j`, `indices(i, epoch)` and `indices(j, epoch)` are disjoint. |
| **Coverage** | Under `Drop`, the union over all ranks equals the dataset minus at most `world_size - 1` trailing samples. |
| **Equal step count** | Under `Pad`, every rank yields exactly `ceil(len / world_size / batch_size)` batches. Ranks that disagree on step count deadlock the collectives. |
| **Cross-rank shuffle consistency** | The permutation is a pure function of `(seed, epoch)`. Ranks derive the same permutation independently and slice different windows of it. No communication. |
| **Identity at world size 1** | `DistributedSampler { world_size: 1, rank: 0, .. }` is behaviorally identical to no sampler. |

The step-count invariant is the one that turns a subtle bug into a hang: if rank 0 runs 100 steps and rank 1 runs 99, rank 0 blocks forever in the 100th all-reduce.

## Background

- **Derive, do not communicate.** The permutation must be `permute(seed, epoch)` computed identically on every rank, with each rank taking the stride-`world_size` slice at its own offset. Any design that broadcasts an index list adds a synchronization point per epoch for no benefit.
- **`set_epoch` is not optional bookkeeping.** Without it the permutation is fixed and every epoch draws the same per-rank partition, which is a well-known way to lose most of the benefit of shuffling. It is also the single most-forgotten call in this API across the ecosystem, which argues for making the epoch a required argument rather than a setter.
- **`Drop` and `Pad` are not interchangeable, and the difference matters for evaluation.** `Pad` double-counts samples, so a validation metric computed under `Pad` is wrong by a small, silent amount. Training tolerates it; evaluation does not. The policy must be explicit at construction rather than defaulted.
- **The worker-backed path is where this gets missed.** `incin-data` supports both zero-worker lazy iteration and worker-backed loading. Sharding applied only to the former is a bug that appears the moment someone sets `num_workers`, and it must be covered by its own test.

## How this interoperates

- **#97** is the consumer, and its correctness depends entirely on this. The data-parallel training test in #97 should assert that ranks receive *different* batches, so this cannot silently regress from that side either.
- **`DistributedContext`** (`crates/incin-core/src/dist/context/`) is the source of `rank` and `world_size`. Reading from it rather than from environment variables directly keeps one identity source and works under the reference transport as well as NCCL.
- **The reference transport** makes this testable in-process: a two-rank in-process run can assert disjointness and coverage without any hardware, so this belongs in `ci.yml` rather than `hardware.yml`.
- **`incin-data`'s typed error contract** (hardened in #43 for 0.1.0) is where an invalid `(rank, world_size)` pair reports: `rank >= world_size` is a construction error, not a panic.
- **#99 (FSDP, tensor, pipeline)** reuses this unchanged for its data-parallel axis. Tensor and pipeline parallelism deliberately do *not* shard data across their axes, and the mesh coordinate is what distinguishes them, so the sampler should take the data-parallel degree from the mesh rather than the world size once #99 lands.

## Acceptance criteria

- With world size N, the union of samples seen by all ranks in one epoch equals the dataset with no overlap, under `RemainderPolicy::Drop`.
- Under `RemainderPolicy::Pad`, every rank yields an identical step count.
- The permutation for a given `(seed, epoch)` is identical across ranks, verified by constructing samplers for all ranks in one process and comparing.
- A sampler with `world_size: 1` is behaviorally identical to today's loader, verified by comparing full epoch sequences.
- Both the zero-worker and worker-backed paths are covered by tests.
- `rank >= world_size` is a typed construction error.
- The `Drop`/`Pad` distinction and its consequence for evaluation metrics is documented in `docs/book/src/data_loading.md`.

## Alternatives

- **Leave sharding to the user**, slicing the dataset before constructing a loader. No API change, and it makes the most common distributed mistake in the ecosystem the default behavior of the library.
- **Infer rank from environment variables inside `DataLoader`.** Removes the explicit call and makes the loader's behavior depend on ambient process state, which is invisible at the call site and untestable in-process.
- **Default to `Pad`.** Guarantees no deadlock; silently corrupts evaluation metrics. Requiring the choice is worth the one line.

## Additional context

Relevant code: `crates/incin-data/src/` (loader and sampler construction, both worker paths), `crates/incin-core/src/dist/context/` (`DistributedContext`, `RANK_ENV`, `WORLD_SIZE_ENV`), `docs/book/src/data_loading.md`.

Prerequisite for #97.

**Comment by xupremix** [comment]
Implemented on `issue/98-rank-aware-sharding`, merged to `develop` (26635caf). Not merged to master, per the 0.2.0 flow: this ships with the milestone.

**What landed**
- `DistributedSampler { world_size, rank, remainder }` + `RemainderPolicy::{Drop, Pad}` in `incin-data`, builder-integrated via `.sampler(...)`
- One shared permutation derived from `(seed, epoch)` per rank; each rank reads only its stride-`world_size` slice &mdash; disjoint by construction, no communication, no second identity source (rank/world size are caller-supplied from the distributed context)
- `set_epoch` rotates partitions across epochs; `world_size: 1` is behaviorally identical to no sampler; `rank >= world_size` and zero world size are typed construction errors (`Error::InvalidModuleState`, matching the existing loader validation pattern)
- Applied at the single index-vector choke point, so the zero-worker and worker-backed paths shard identically

**Acceptance criteria vs tests** (`crates/incin-data/tests/distributed_sampler.rs`, 7 passing)
- Disjointness + coverage under `Drop` (union = LEN − LEN mod W): ✅
- Identical step counts under `Pad`: ✅
- Cross-rank permutation consistency: ✅ verified by interleaving rank slices back into the exact unsharded-loader order for the same `(seed, epoch)`
- World-size-1 identity vs today's loader: ✅
- Both worker paths covered and proven equivalent: ✅
- Typed construction error: ✅
- Book documentation of `Drop`/`Pad` and the evaluation-metric caveat (`data_loading.md` § Data-parallel sampling): ✅

Checks run: full `incin-data` suite (42 passing), clippy `-p incin-data --all-targets` clean, facade check, docs checks + book build.

#97 can now assume disjoint per-rank batches; its data-parallel test should still assert it end to end. Note for #99: when mesh axes arrive here, the sampler's degree should come from the data-parallel axis rather than world size.
